Debug documentation (#11)
balancer/balancer-v3-erc4626-tests: Readme.md

- `withdraw`
- `redeem`

### Debug failing tests

> There are some common errors that may occur when testing a token. The step-by-step below helps to cover most of them.

Reviewer: ```suggestion
There are some common errors that occur when testing a token. The step-by-step guide below helps to cover most of them.
```

> 1. Get the wrapped token address, go to etherscan;

Reviewer: ```suggestion
1. Get the wrapped token address from Etherscan.
```
I don't think we need all the semi-colons.

Also not sure I follow here. Do you mean navigate to the wrapped token on Etherscan? (Seems not, since the instructions below interact with the Vault.) Does this mean get the token address from the pool (e.g., IPoolInfo) or the Vault (e.g., getPoolTokens)?

Author: It means to navigate to wrapped token on Etherscan. I'll rewrite it.

Reviewer: Ok; just "Navigate to the wrapped token address on Etherscan."

> 2. Check if the vault's buffer, in the current block, has liquidity for that token (Use vault explorer's `getBufferBalance` using the wrapped token address as argument to check the liquidity)

Reviewer: ```suggestion
2. Check whether the Vault's buffer at the current block has liquidity for that token (Use the Vault Explorer's `getBufferBalance` function with the wrapped token address as the argument.)
```

>     1. If there is, it's probably reverting because the buffer is already initialized. Choose a block number prior to the buffer initialization and try again;

Reviewer: ```suggestion
    1. If there is, it's probably reverting because the buffer was already initialized. Choose a block number prior to the buffer initialization and try again.
```

>         1. If the wrapper was recently created, put a block number right after the creation of the wrapper and test again;

Reviewer: ```suggestion
        1. If the wrapper was recently created, use a block number right after the creation of the wrapper and try again.
```

>         2. If not, try to put an old block number, closer to Balancer V3 release, where the vault is already there but the buffer is not initialized;

Reviewer: ```suggestion
        2. If not, try to put an old block number, closer to the Balancer V3 launch (e.g., 21332121 on Mainnet), where the Vault is already there but the buffer was not yet initialized.
```

3. Check the asset of the wrapped token (underlying token) and get the holders;

>     1. Check if the top holder is the one used in the test;

Reviewer: ```suggestion
    1. Check whether the top holder is the one used in the test;
```

>         1. If it's not, change the holder to the top one (it may be a bit tricky if we're using an old block number, then you probably need to iterate over the holders list until you find a holder with balance in that specific block number)

Reviewer: ```suggestion
        1. If it's not, change the holder to the top one. It may be a bit tricky if you're using an old block number. In that case, you would need to iterate over the holders list until you find a holder with sufficient balance at that specific block.
```

        2. Check if the holder we're using has enough balance to cover 3 * defaultAmount. It means, if default amount is 1e6 * 1e18, the holder must have at least 3e6 * 1e18 assets;

> 4. Finally, if the error persists, it needs to be tested in [balancer-v3-monorepo](https://github.com/balancer/balancer-v3-monorepo);

Reviewer: ```suggestion
4. Finally, if the error persists, try testing it in the [balancer-v3-monorepo](https://github.com/balancer/balancer-v3-monorepo).
```

>    1. In that project, go to `pkg/vault/test/foundry/fork` and copy the test `ERC4626MainnetAaveUsdc.t.sol` (or modify it);

Reviewer: ```suggestion
   1. In your copy of that repo, go to `pkg/vault/test/foundry/fork`, then copy and modify the `ERC4626MainnetAaveUsdc.t.sol` test for your token.
```

>    2. Run the test specifically for this file (something like `yarn test:forge --match-contract ERC4626MainnetAaveUsdcTest` inside the vault pkg folder) and check the error;

Reviewer: ```suggestion
   2. Run the tests specific to this file (something like `yarn test:forge --match-contract ERC4626MainnetAaveUsdcTest` inside the pkg/vault folder) and check for errors.
```

>    3. If the test passes, it means that the buffer is initialized in the chosen block number (the test in the monorepo uses a clean vault, and this is the only reason for a test to pass in the monorepo and don't pass in this ERC4626 tests repo);

Reviewer: ```suggestion
   3. If the tests pass, it means that the buffer is initialized at the chosen block number. The test in the monorepo uses a newly deployed Vault: this is the only reason for a test to pass in the monorepo that doesn't pass in the ERC4626 tests repo.
```

> If the step-by-step above does not help, probably the token is incompatible with the vault and need further investigation.

Reviewer: ```suggestion
If the step-by-step instructions above do not help, the token is likely incompatible with the vault, and needs further investigation.
```
